Primitive spreadsheet

donboie opened the pull request.

danieldresser committed:
NodeGraph : Use viewport rasterToWorldSpace

avoids of relying on specifics of camera support in viewportGadget

danieldresser committed:
MapProjection : Adapt to new camera interface

danieldresser committed:
LightToCamera : Adapt to new Camera interface

danieldresser committed:
CameraVisualiser : Adapt to new camera interface

danieldresser committed:
GafferScene::Camera : Adapt to new Cortex camera interface

danieldresser committed:
IECoreAppleseedPreview::Render : Adapt to new camera interface

danieldresser committed:
IECoreArnoldPreview::Renderer : Adapt to new camera interface

danieldresser committed:
ViewportGadget : Adapt to new camera interface

danieldresser committed:
SceneView : Adapt to new camera interface

danieldresser committed:
IECoreGLPreview : Adapt to new camera interface

dboogert committed:
GafferSceneUI : PrimitiveInspector

danieldresser committed:
StandardOptions : Added filmFitMode render option

danieldresser committed:
Add CameraTweaks node

andrewkaufman commented:
@donboie are you wanting code comments yet or mainly playing with the ui myself and reactions to look/feel?

dboogert commented:
> @donboie are you wanting code comments yet or mainly playing with the ui myself and reactions to look/feel?

Mainly the later

andrewkaufman commented:
- I'd like these for first release
  - [x] I wish the Node/Location info at the top looked like it does in the _Scene Inspector_ (without the diff part for now I guess)
  - [x] I find it awkward to have it on the same pane as the Viewer. I can't look at the mesh and inspect its prim vars at the same time.
  - [x] Already mentioned on the ticket, but I really want to see Interpretation and Type with each prim var.
  - [ ] The columns on a given tab should be alphabetically sorted
  - [x] Can we add some contrast between prim var columns? The _Scene Inspector_ and _Scene Hierarchy_ use alternating light gray / dark gray rows. Maybe we can use those same style hints? This might help to identify the boundary between prim vars vs x/y/z components.
  - Testing on a production ParticleGeo, I got an error that prevents Vertex prim vars from drawing at all. Its from an unsupported prim var type (orient `QuatfVectorData`).
    - [x]  We should support quats for this release. It'll come up in production immediately.
    - [x]  We should also gracefully handle non-supported types somehow.
      - Maybe display the other prim vars and list the heading for the broken one but with an error widget instead of its column?
      - It'd also be great for the actual traceback to list the bad prim var name. Currently its just `assert( accessor is not None )`

- Would be nice, but maybe can wait on these, unless they're trivial:
  - [ ]  I'd love a filter to reduce the number of columns displayed
  - [ ]  It'd be nice to add a menu item to the gear menu in the _Scene Inspector_ for the prim var. `Gear->Inspect` could reveal (or acquire) the _Primitive Inspector_ and also reveal the proper interpolation tab.
  - [ ]  Might be nice to have a summary (num elements, num unique elements) but that should be for the whole prim var not the individual x/y/z components.
  - [ ]  Its a bit unexpected that nothing in this pane is drag/droppable. Most panes have something... I'm not sure what it should be though... maybe the selected cell values?

andrewkaufman commented:
>  The columns on a given tab should be alphabetically sorted

Circling back on this, alphabetic probably isn't what we want. P should be the first one in Vertex regardless. We probably want to prioritize the ordering for P, Pref, N, & uv, since they all have special meaning. I first suggested alphabetic because of a production point cloud I was inspecting. It displayed Constant prim vars in this order: varmap (string), boxMin (v3f), sceneFile (string), boxMax (v3f). I found it odd that boxMin and boxMax weren't listed next to each other...

Thoughts?

donboie commented:
@andrewkaufman would you take another look?

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -95,11 +95,13 @@ def __init__(
+
 		self.__tableView.verticalHeader().setVisible( showIndices )
-		QtCompat.setSectionResizeMode( self.__tableView.verticalHeader(), QtWidgets.QHeaderView.Fixed )
+		QtCompat.setSectionResizeMode( self.__tableView.verticalHeader(), QtWidgets.QHeaderView.Stretch )
```
What's the reason for this? It results in unnecessarily stretched rows, as below :

![stretchedcolumns](https://user-images.githubusercontent.com/1133871/42994564-5740424a-8c06-11e8-8171-bfe93fdfc8aa.png)

Reverting this line got things back to what I would expect.

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -95,11 +95,13 @@ def __init__(
 		self.__tableView.horizontalHeader().setVisible( bool( header ) )
 		self.__tableView.horizontalHeader().setMinimumSectionSize( 70 )
 
+		QtCompat.setSectionResizeMode( self.__tableView.horizontalHeader(), QtWidgets.QHeaderView.Stretch )
```
After playing with it both ways, I'm not sure I know what difference this is making. I'm concerned about knock-on effects to existing UIs though. Do we need it?

johnhaddon replied:
Hang on, aren't we changing this again in `setData()` anyway?

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -95,11 +95,13 @@ def __init__(
 		self.__tableView.verticalHeader().setObjectName( "vectorDataWidgetVerticalHeader" )
 
-		self.__tableView.setHorizontalScrollBarPolicy( QtCore.Qt.ScrollBarAlwaysOff )
+		self.__tableView.setHorizontalScrollBarPolicy( QtCore.Qt.ScrollBarAsNeeded )
```
We probably shouldn't change the default for fear of breaking existing UIs. I suggest we just add public `setHorizontalScrollMode()` and `getHorizontalScrollMode()` methods. We already have enums for this in ScrolledContainer, so we should probably port them out of there and into GafferUI.Enums so they can be reused. Let's do this is a single commit on its own.

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+
+			if self.__outputScenePlugs and len( self.__outputScenePlugs ) :
+
+				self.__nodeLabel.setText( self.__outputScenePlugs[-1].node().getName() )
```
This doesn't update correctly when the node name changes. Use a NameLabel widget and call `label.setGraphComponent( node )` and it will do it for you.

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+					self.__locationFrame._repolish()
+
+					obj = self.__outputScenePlugs[-1].object( self.__targetPaths[-1] )
+					if obj and obj != IECore.NullObject() :
```
Should be `if isinstance( obj, IECoreScene.Primitive )`, otherwise things like this happen :

```
ERROR : EventLoop.__qtIdleCallback : Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/python/GafferUI/EventLoop.py", line 275, in __qtIdleCallback
    if not c() :
  File "/disk1/john/dev/build/gaffer/python/GafferUI/LazyMethod.py", line 161, in __idle
    cls.ERROR : EventLoop.__qtIdleCallback : Traceback (most recent call last):
  File "/disk1/john/dev/build/gaffer/python/GafferUI/EventLoop.py", line 275, in __qtIdleCallback
    if not c() :
  File "/disk1/john/dev/build/gaffer/python/GafferUI/LazyMethod.py", line 161, in __idle
    cls.__doPendingCalls( widget, method )
  File "/disk1/john/dev/build/gaffer/python/GafferUI/LazyMethod.py", line 174, in __doPendingCalls
    method( widget, *pendingCall.args, **pendingCall.kw )
  File "/disk1/john/dev/build/gaffer/python/GafferSceneUI/PrimitiveInspector.py", line 263, in __updateLazily
    self.__update()
  File "/disk1/john/dev/build/gaffer/python/GafferSceneUI/PrimitiveInspector.py", line 303, in __update
    for primvarName in _orderPrimitiveVariables( obj.keys() ) :
AttributeError: 'Camera' object has no attribute 'keys'__doPendingCalls( widget, method )
  File "/disk1/john/dev/build/gaffer/python/GafferUI/LazyMethod.py", line 174, in __doPendingCalls
    method( widget, *pendingCall.args, **pendingCall.kw )
  File "/disk1/john/dev/build/gaffer/python/GafferSceneUI/PrimitiveInspector.py", line 263, in __updateLazily
    self.__update()
  File "/disk1/john/dev/build/gaffer/python/GafferSceneUI/PrimitiveInspector.py", line 303, in __update
    for primvarName in _orderPrimitiveVariables( obj.keys() ) :
AttributeError: 'Camera' object has no attribute 'keys'
```

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+				self.__plugDirtiedConnections.append( node.plugDirtiedSignal().connect( Gaffer.WeakMethod( self.__plugDirtied ) ) )
+				self.__parentChangedConnections.append( outputScenePlugs[0].parentChangedSignal().connect( Gaffer.WeakMethod( self.__plugParentChanged ) ) )
+
+			self.__outputScenePlugs = outputScenePlugs
```
Unlike the SceneInspector, this only supports a single scene, so let's just store a single scene instead of a list. It's misleading otherwise...

andrewkaufman replied:
Won't it need to support multiple scenes in the future, when we can diff primitives?

johnhaddon replied:
Sure, but we're a long ways away from that...

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+
+		self.__updateLazily()
+
+	def _updateFromContext( self, modifiedItems ) :
```
This isn't updating properly when time changes. See `SceneInspector._updateFromContext()` for a correct example.

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+
+	def __plugDirtied( self, plug ) :
+
+		if isinstance( plug, GafferScene.ScenePlug ) and plug.direction() == Gaffer.Plug.Direction.Out :
```
This is updating when any child of the ScenePlug is dirtied (transform, attributes etc), and also isn't considering if the scene in question is the one we're looking at. We should limit it to only update when it's the "object" plug of the correct ScenePlug...

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+		self.__update()
+
+	def __update( self ) :
+		# The SceneInspector's internal context is not necessarily bound at this point, which can lead to errors
```
Comment is inaccurate/irrelevant - remove.

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -256,6 +259,10 @@ def getData( self ) :
 
+	def setColumnToolTips( self, columnToolTips ) :
+
+		self.__columnToolTips = columnToolTips
```
Shouldn't this update the model? Since the tooltips are intrinsically linked with the number of columns, perhaps the data and tooltips should all be settable with a single call somehow anyway?

And as I've said before, `set` should be paired with `get`.

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -811,20 +818,26 @@ def __init__( self, data, parent=None, editable=True, header=None, columnToolTip
 			assert( accessor is not None )
 			for i in range( 0, accessor.numColumns() ) :
-				self.__columns.append( IECore.Struct( accessor=accessor, relativeColumnIndex=i ) )
+				self.__columns.append( IECore.Struct( accessor=accessor, relativeColumnIndex=i, index = index ) )
```
`index` is the original column index as far as `data` is concerned, right? We have some jiggery pokery for going backwards and forwards on this in `VectorDataWidget.columnToDataIndex()` and `VectorDataWidget.dataToColumnIndex()`. I wonder if we should avoid the redundancy one way or another?

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -905,11 +919,18 @@ def flags( self, index ) :
+		column = self.__columns[index.column()]
+		if role == QtCore.Qt.BackgroundColorRole :
+
+			if column.index % 2 == 0:
```
I'm bikeshedding a bit here, but could these match the colours in the PathListingWidget (HierarchyView). In this context the lighter colour feels like subtle highlighting to me, rather than an alternate background. Maybe we should also update the SceneInspector to match?

donboie replied:
How about updating _Stylesheet to allow querying the parameters which are used to build the stylesheet colours?

johnhaddon replied:
What I'd really like to do longer term is find a way of unifying the `GafferUI::Style` stuff we use for the Gadgets with the stylesheet stuff we use for the Widgets. That's not going to happen now though. Are you thinking of just exposing variables like `_Stylesheet._highlightColor` and so on? As long is it all remains protected that sounds good to me.

johnhaddon commented on python/GafferUI/VectorDataWidget.py:
```diff
@@ -204,8 +206,9 @@ def setData( self, data ) :
 		if data is not None :
-			if not isinstance( data, list ) :
-				data = [ data ]
+			if not isinstance( data, dict ) :
```
This is obviously gonna break every single other user of VectorDataWidget. Looks like the main motivation is to get the headings you want? If so, it is already possible to pass headings as a list via the `header` constructor parameter. Let's go one way or the other, and make sure we don't break other code.

As I noted on `setColumnToolTips()`, it is a bit rubbish that different aspects of the model are specified in different places, so they can get out of sync. I do like the fact that a dictionary keeps the header and columns in a single structure, and had I done that from the start that might have been better. I'm not sure how we'd extend the dict approach to the tooltips as well though, without introducing a public Column class of some sort.

Thoughts? I think it's either charge forwards and switch things up across the board, or use the `header` stuff we already have (with any tweaks needed).

donboie replied:
> This is obviously gonna break every single other user of VectorDataWidget. Looks like the main motivation is to get the headings you want? If so, it is already possible to pass headings as a list via the header constructor parameter. Let's go one way or the other, and make sure we don't break other code.

As you're happy with the idea of this widget being used for it's existing purposes and updated for the PrimitiveInspector's needs then I'll try using the header ctor parameter and see how it looks.

johnhaddon commented on startup/gui/layouts.py:
```diff
@@ -49,9 +49,81 @@
+		)
+	)
+
+def getSource(fn):
```
I don't think defining an unused function only to suck the source out of it is an improvement. The reformatting is an improvement if we want to edit these layouts by hand, but I think what I've actually been doing in practice is modifying them in the UI and then saving them. Let's either revert it all, or keep the formatting but in a simple triple-quoted string.

donboie replied:
> The reformatting is an improvement if we want to edit these layouts by hand, but I think what I've actually been doing in practice is modifying them in the UI and then saving them.

Ah! This explains the strange situation of python code in the python file as a mushed together in a string. I'll revert and add a few comments for future visitors to this file.

johnhaddon commented on python/GafferSceneUI/PrimitiveInspector.py:
```diff
@@ -0,0 +1,321 @@
+
+def conditionPrimvar( primvar ) :
+
+	scalarToVector = {
```
I think something like this might be a shorter alternative :

```
if hasattr( data, "value" ) :
    data = ECore.DataTraits.dataFromElement( [ data.value ] )
```

johnhaddon reviewed (commented):
Nice Don!

I haven't been able to resist the temptation to review more than just the VectorDataWidget changes, since I think we're getting close and there are a few obvious problems that will need fixing anyway.

As for the VectorDataWidget, I've made a few comments inline, but to me these changes look like they will break things pretty badly (just try making a PathFilter). I think some of the stuff you're doing highlights problems with the original API, but we need to find a way of doing what we need _and_ preserving backwards compatibility. Have a read of my comments and we'll figure something out...

Cheers...
John

donboie commented:
> I haven't been able to resist the temptation to review more than just the VectorDataWidget changes, since I think we're getting close and there are a few obvious problems that will need fixing anyway.

Thanks for taking the time!

> As for the VectorDataWidget, I've made a few comments inline, but to me these changes look like they will break things pretty badly (just try making a PathFilter). I think some of the stuff you're doing highlights problems with the original API, but we need to find a way of doing what we need and preserving backwards compatibility. Have a read of my comments and we'll figure something out...

I suppose I was considering if it was worth making the VectorDataWidget work for all cases or how the potentially different behaviour should be factored.

donboie committed:
GafferUI : Move ScrollMode enum from ScrolledContainer to Enums

donboie committed:
GafferUI : Stylesheet - allow query of style colors.

mattigruener committed:
AnimationGadget : remove k-drag and allow dragging from the time axis instead

mattigruener committed:
ViewportGadget : support drag tracking for individual axes

mattigruener committed:
AnimationGadget/AnimationEditor : move dragTracking setting to AnimtationGadget

And disable dragging in Y when moving the frame indicator.

mattigruener committed:
GraphEditor : adjust to new drag tracking options

mattigruener committed:
ViewportGadget : expose drag tracking options in python

danieldresser committed:
RendererController : Update for new camera interface

johnhaddon committed:
ShadingEngine : Support more vector type conversions

Among other things, this allows us to read Color4fData as `color` values in OSLObject, by discarding the alpha channel.

danieldresser committed:
SceneAlgo : Adapt to new camera interface

danieldresser committed:
TestCase : Keep reference plug values during assertHashesValid

danieldresser committed:
GafferScene::Camera : Update tests

danieldresser committed:
IECoreGLPreviewTest : Use exr for temp image files.

Tiff should be fine, but at IE, we seem to have some sort of config that assumes tiffs are in a weird color space and screws up the test.  We can get around this by using an isolated env, but since most of our test files use exrs anyway, it seems more reliable and consistent to always use exr.

johnhaddon committed:
TransformTool : Keyframe transform plugs when "S" is pressed

Commit:
PathListingWidget : only initiate drag if valid list entry under pointer

johnhaddon committed:
TransformTools : Support animation

Fixes #2721

johnhaddon committed:
TransformTool : Provide control over the size of the handles

Fixes #2671

Breaking Changes :

- TransformTool : Added argument to `updateHandles()` method.

johnhaddon committed:
Merge branch 'catalogueMultiSelect' of https://github.com/mattigruener/gaffer into mattigruener-catalogueMultiSelect

johnhaddon committed:
CatalogueUI : Remove drag pointer from image listing

This is in keeping with the other drag-to-reorder UIs we have in the UIEditor.

johnhaddon committed:
Merge branch 'mattigruener-catalogueMultiSelect'

Manual merge of #2702

johnhaddon committed:
Merge pull request #2623 from medubelko/docPerformanceEdit

Docs: Edit/move "Managing Complexity" → "Script Performance"

Commit:
AnimationGadget : replace range-for with regular for to fix bug

danieldresser committed:
SceneView : Update tests for new cameras

danieldresser committed:
GafferImage::Offset : Fix bug that could trigger hang

johnhaddon committed:
Merge pull request #2770 from danieldresser-ie/offsetFix

GafferImage::Offset : Fix bug that could trigger hang

johnhaddon committed:
DependencyNodeBinding : Guard against `None` from `affects()` overrides

We don't want coding errors in Python to be able to take down the process.

johnhaddon committed:
ParallelAlgoTest : Fix bugs in ExpectedUIThreadCall

- Protect `__conditionStack` access with mutex. Without this, we could trigger the top condition more than once, and fail to trigger other conditions. This lead to hangs in `__exit__()`, particularly in CatalogueTest on Travis.
- Only disconnect from the signal when the stack is empty.

andrewkaufman committed:
Merge pull request #2624 from medubelko/docInstallationEdit

Docs: Edit/move "Installation" → "Getting Started"

andrewkaufman committed:
Merge pull request #2767 from johnhaddon/catalogueDeleteFix

GraphComponent : Restore original order when undoing `removeChild()`

andrewkaufman committed:
Merge pull request #2771 from johnhaddon/dependencyNodeBindingFix

DependencyNodeBinding : Guard against `None` from `affects()` overrides

andrewkaufman committed:
Merge pull request #2772 from johnhaddon/catalogueHangTesting

ParallelAlgoTest : Fix bugs in ExpectedUIThreadCall

andrewkaufman committed:
Merge pull request #2769 from johnhaddon/transformToolAnimation

TransformTool improvements

Commit:
SceneInspector : add inspection of shader parameters

NOTE: only the first shader in a network is currently supported.

These should work:

- lights
- light modulators
- surface shaders

Commit:
StandardGraphLayout : no need for auxiliary member on Edges

johnhaddon committed:
Merge pull request #2779 from mattigruener/auxLayoutCleanup

StandardGraphLayout : no need for auxiliary member on Edges

johnhaddon committed:
Merge pull request #2768 from mattigruener/animationEditorPrecisionIssues

AnimationEditor precision issues

johnhaddon committed:
ScriptNodeBinding : Fix GIL management bugs

Now we have asynchronous compute, it is critical that any graph editing operation releases the GIL. This is because the edit will cancel all background tasks and wait for them before proceeding. If the background task is implemented in Python, it will need to acquire the GIL before it can be cancelled, therefore the edit must release the GIL first.

These bugs first manifested in the wild as hangs when cutting and pasting nodes.

andrewkaufman committed:
Merge pull request #2780 from johnhaddon/missingGILReleases

ScriptNodeBinding : Fix GIL management bugs

Commit:
Beginner Tutorial : Update link in home page

andrewkaufman committed:
Merge pull request #2629 from medubelko/docsBeginnerTutorial

Docs: Edit/rename  "Getting Started" tutorial → "Tutorial: Assembling the Gaffer Bot"

johnhaddon committed:
ParallelAlgo : Refactor `callOnUIThread()` internals

This has been working perfectly well in the real world, but continuing to cause problems when running the Catalogue unit tests on Travis.

The issue was related to the lack of thread safety in `CallOnUIThreadSignal` and the `ParallelAlgoTest.ExpectedUIThreadCall` we use to emulate the UI for testing purposes. The `ExpectedUIThreadCall` class was sometimes modifying its connection while the signal was still being emitted, with the most common symptom being that a slot was called twice (once by the old connection, once by the new one), or the `CallOnUIThreadSignal` block exited before all expected work was done. This could manifest as either hangs or test failures.

Since using a signal didn't make that much sense anyway (we only want one thing connected at any time) I've replaced it with a simple `registerUIThreadCallHandler()` function, which allows us to manage an internal mutex around the handler.

> Tip : For a long time these problems never manifested locally, only on Travis.
> Using `taskset 0x01 gaffer test ...` made them appear locally too, as it forces
> Gaffer onto a single CPU, which I presume yields threading patterns closer to
> those seen on Travis.

johnhaddon committed:
FilteredSceneProcessorUI : Fix bug in "Select Affected" menu item

It didn't accound for the fact that some FilteredSceneProcessors (CopyAttributes for instance) have ArrayPlug inputs.

johnhaddon committed:
SetUI : Recognise ScenePlugs inside ArrayPlugs

This makes the SetFilter set menu items work with the CopyAttributes node.

johnhaddon committed:
Travis : Update to GafferHQ/dependencies version 0.49.0.0

Commit:
AnimationGadget : use new drag tracking when moving keys in just one axis

donboie committed:
GafferUI : VectorDataWidget - Use GafferUI.ScrollMode

johnhaddon committed:
Merge pull request #2782 from johnhaddon/releasePrep49

Travis : Update to GafferHQ/dependencies version 0.49.0.0

johnhaddon committed:
Bump version to 0.49.0.0 and update Changes

johnhaddon committed:
Merge pull request #2759 from johnhaddon/oslColor4f

ShadingEngine : Support more vector type conversions

johnhaddon committed:
Merge branch 'frameIndicatorUpdate' of https://github.com/mattigruener/gaffer into mattigruener-frameIndicatorUpdate

Manual merge of PR #2724, to fix conflicts in AnimationGadget.cpp.

johnhaddon committed:
Merge branch 'mattigruener-frameIndicatorUpdate'

donboie committed:
GafferUI : VectorDataWidget - Matrix Accessor

register a DataAccessor for M33f, M44f, M33d & M44d

donboie committed:
GafferUI : VectorDataWidget - Box Accessor

Register Data accessors for Box2i, Box2f, Box2d,  Box3i, Box3f & Box3d

donboie committed:
GafferUI : VectorDataWidget - Registered additional vector types

V2i, V2f, V2d, V3i & V3d

donboie committed:
GafferUI : VectorDataWidget - Quaternion Accessor

donboie committed:
GafferUI : VectorDataWidget - Registered delegates for new types

donboie committed:
GafferUI : VectorDataWidget - Registered additional integral  types

u8, s8, u16 & s16

donboie committed:
GafferUI : VectorDataWidget - Refactored column index functions

Move columnToDataIndex & dataToColumnIndex into _Model class. Hopefully these can be used to set alternate background colors for the columns.

donboie committed:
GafferUI : VectorDataWidget - Added option to change stretch behaviour

donboie committed:
GafferUI : VectorDataWidget - Added option for tooltips behaviour

- defines if the tooltip be displayed over the data region of the Widget or not (in addition to the column heading)

donboie committed:
GafferUI : VectorDataWidget - Support optional column headings

donboie committed:
GafferUI : VectorDataWidget - Alternate column background colour

donboie committed:
GafferUI : VectorDataWidget - Accessors for headers & tooltips

Commit:
"Anatomy of a Scene" : Add screenshots

Commit:
"Anatomy of a Scene" : Add article outlining data in a 3D scene

Commit:
"Working with Scenes" : Add section for 3D scenes

donboie committed:
GafferUI : VectorDataWidget - Fixed getData when model not set.

johnhaddon committed:
GraphGadget bindings : Release GIL for all Action-triggering methods

johnhaddon committed:
BackdropNodeGadget binding : Release GIL in `frame()` method

johnhaddon committed:
Switch binding : Release GIL in `setup()`

johnhaddon committed:
SplinePlug binding : Release GIL in all Action-triggering methods

johnhaddon committed:
PlugAlgo binding : Release GIL in all Action-triggering methods

johnhaddon committed:
PlugBinding : Release GIL for all Action-triggering methods

johnhaddon committed:
Metadata binding : Release GIL for `deregisterValue()`

johnhaddon committed:
GraphComponent binding : Release GIL in all Action-triggering methods

johnhaddon committed:
Dot binding : Release GIL in `setup()`

johnhaddon committed:
BoxPlug binding : Release GIL in `setValue()`

johnhaddon committed:
BoxIO binding : Bind `setup()`

johnhaddon committed:
MetadataAlgo binding : Release GIL for Action-triggering methods

johnhaddon committed:
SceneReader : Add transform plug

This makes it possible to position a cache without needing a separate Transform node, so SceneReaders can be used more like Light and ObjectSource nodes.

johnhaddon committed:
TransformTool : Support SceneReader transforms

johnhaddon committed:
FileMenu : Ask for confirmation in "Revert to Saved"

Also properly report any errors that occur during loading.

Fixes #2735

Commit:
"Scripting Performance" : Fix BuildTarget of screenshots

Commit:
"Beginner Tutorial" : Improve copy regarding scene paradigm

Commit:
"Beginner Tutorial" : Add links to "Getting Started" chapter articles

Commit:
"Beginner Tutorial" : Swap left/right leg in scene traversal section

Commit:
"Beginner Tutorial" : Screengrab cleanup

Commit:
"Beginner Tutorial" : Adjust camera position to centre Gaffy in frame

danieldresser committed:
SceneView : Remove internal processing node that appears unnecessary

johnhaddon committed:
Merge pull request #2789 from johnhaddon/gilReleaseAudit

GIL management audit

johnhaddon committed:
Bump version to 0.49.0.1 and update Changes

johnhaddon committed:
Merge pull request #2787 from medubelko/docsSceneAnatomy

Docs : Add "Anatomy of a Scene"

johnhaddon committed:
SubGraph : Fix crash bug in `correspondingInput()`

This typically manifested itself as a crash when trying to delete or cut a box.

Commit:
"Beginner Tutorial" : Fix a screenshot path

danieldresser committed:
IECoreGLPreviewRender : Initialize GL before batchRender

Fixes crash when trying to use `gaffer execute` to run GL renders

andrewkaufman committed:
Merge pull request #2796 from johnhaddon/subgraphFix

SubGraph : Fix crash bug in `correspondingInput()`

andrewkaufman committed:
Merge pull request #2794 from johnhaddon/revertConfirmation

FileMenu : Ask for confirmation in "Revert to Saved"

andrewkaufman committed:
Merge pull request #2792 from johnhaddon/sceneReaderTransform

 SceneReader : Add transform plug

danieldresser committed:
IECoreGLPreview::Renderer : Don't render the visualiser of current cam

andrewkaufman committed:
Merge pull request #2785 from johnhaddon/copyAttributesUIFix

CopyAttributes UI fixes

johnhaddon committed:
MatchPatternPathFilter : Make exception-safe

`std::remove_if()` has undefined behaviour if an exception is thrown, and in practice we end up with a child vector containing null pointers. This was causing the file browser to crash if pointed at files with strange permissions.

The fix is to catch the exception and turn it into an error. The alternatives would be to allow the exception to be thrown but to use something more robust than `remove_if()`. This doesn't seem at all useful to client code : it wants to be able to show what it can, rather than allow a single bad file to prevent use listing all the others.

Fixes #2798.

johnhaddon committed:
Merge pull request #2795 from medubelko/docsBeginnerTutorialUpdate

Docs : "Beginner Tutorial" update

johnhaddon committed:
BackdropNodeGadget : Fix GIL management for `frame()` method

`extend_container()` uses the Python API, so we can't release the GIL until after we call it.

johnhaddon committed:
Bump version to 0.49.1.0 and update Changes

johnhaddon committed:
TransformTool : Move `orientedTransform()` method to `Selection` class

Breaking Change :

- TransformTool : Moved `orientedTransform()` method to `Selection` class.

johnhaddon committed:
TransformTools : Support multiple selection

Fixes #2665.

Breaking Change :

- TransformTool : Changed `selection()` return type.

johnhaddon committed:
TransformTool : Fix bug when same node is source of multiple transforms

andrewkaufman committed:
Merge pull request #2801 from johnhaddon/moreGILManagementWoes

BackdropNodeGadget : Fix GIL management for `frame()` method

andrewkaufman committed:
Merge pull request #2800 from johnhaddon/fileBrowserCrashFix

MatchPatternPathFilter : Make exception-safe

andrewkaufman committed:
Merge pull request #2786 from johnhaddon/catalogueTestFailures

ParallelAlgo : Refactor `callOnUIThread()` internals

andrewkaufman committed:
Update changes

johnhaddon committed:
ContextAlgoTest : Remove unused code

johnhaddon committed:
GafferSceneUI::ContextAlgo : Add `set/getLastSelectedPath()` functions

And associated `affectsLastSelectedPath()` function.

johnhaddon committed:
SelectionTool : Use `ContextAlgo::setLastSelectedPath()`

johnhaddon committed:
TransformTool : Display handles for last selected object

Commit:
SceneInspector : fix formatting for imath.Color4f data type

johnhaddon committed:
Gaffer : Remove SwitchDependencyNode

It is not used anywhere, and removing it allows us to remove some complexity. This is a first step towards detemplating the Switch class and removing all the specialised Switch nodes.

johnhaddon committed:
Config : Remove InputGenerator compatibility

This dates back to version 0.16.0.0 (August 2015) when we moved a bunch of nodes from using this nasty old InputGenerator thing to using the ArrayPlug that we use today. Removing the compatibility means that files from version 0.15.0.0 will no longer load - they will need to be resaved from a version after 0.16.0.0 and before this commit is merged.

This is a prerequisite for embarking on #2123. Strictly speaking I could probably do some fiddling to keep this compatibility in place, but given the age of the version in question I don't think it's worth the effort.

Breaking Changes :

- Removed compatibility for loading files from Gaffer 0.15.0.0 and earlier. Resave the file
  in a more recent version before loading in the current version. Note that this can also expose
  bugs in custom scripts : if a script incorrectly attempts to connect to an ArrayPlug instead
  this will now fail instead of being silently corrected.

donboie committed:
GafferUI : VectorDataWidget - Document headerPrefix vs header parameter.

johnhaddon committed:
Merge pull request #2803 from johnhaddon/transformToolMultiSelect

TransformTool multi select

johnhaddon committed:
CameraTool : Avoid unnecessary updates

johnhaddon committed:
TransformTool : Mitigate against Monitor scoping issues

These can occur when using a TransformTool while the Catalogue is saving an image on a background thread.

Fixes #2807.

johnhaddon committed:
Config : Remove filterCompatibility.py

This belongs to the same era as the InputGenerator compatibility files removed in #2805.

> Note : IE developers should be aware that some internal code is still referring to
> plugs named "match", and this will need updating to "out".

johnhaddon committed:
Switch : Add `inPlugs()` and `outPlug()` methods

andrewkaufman committed:
Merge pull request #2808 from johnhaddon/cameraToolFixes

CameraTool crash bandaid

johnhaddon committed:
Bump version to 0.50.0.0 and update Changes

johnhaddon committed:
ContextAlgo : Add GlobalScope utility class

johnhaddon committed:
Switch : Replace SwitchTraits with ContextAlgo::GlobalScope

This means that a regular SwitchComputeNode now has the same context optimisations as the specialised SceneSwitch and ImageSwitch nodes.

johnhaddon committed:
GafferScene : Remove SceneSwitch

Breaking Changes :

- GafferScene : Removed SceneSwitch node : use a regular Switch instead. Backwards compatibility
  for old files is provided by converting SceneSwitches during loading.

johnhaddon committed:
Python files : Remove reliance on switch compatibility configs

There is some subtlety here for the ShaderSwitch case. When ShaderSwitches were first introduced, they always had "in" and "out" plugs added during construction, but in 0.31.0.0 this ceased to be the case, so that we could support multiple parameter types (previously we only supported RSL coshaders). When we did that, we added `shaderSwitchCompatibility.py` to auto-add the old "in"/"out" plugs on first attempted access, and we still have that same compatibility code in our new `switchCompatibility.py` file. The difference is that the new ShaderSwitch stub isn't a true C++ class as before, so the compatibility `__getitem__` is lost through slicing when retrieving the node from C++. The bottom line is this :

```
s = GafferScene.ShaderSwitch()
s["in"] # Works

script = Gaffer.ScriptNode()
script["s"] = GafferScene.ShaderSwitch()
script["s"]["in"] # Fails
```

The compatibility we have now is sufficient for loading old files, but will fail for any custom code which has the latter form above. The existence of such code seems unlikely since to support current shader parameters (rather than RSL coshaders) a call to `setup()` is required. But I felt it was important to document this in case...

johnhaddon committed:
GafferScene : Remove ShaderSwitch

Breaking Changes :

- GafferScene : Removed ShaderSwitch node : use a regular Switch instead. Backwards compatibility
  for old files is provided by converting ShaderSwitches during loading.

johnhaddon committed:
GafferScene : Remove FilterSwitch

Note the special handling required in `FilterPlug::sceneAffectsMatch()` to keep the query working when a FilterPlug receives an input from a context-varying Switch. This is similar to the custom handling required in Shader/ShaderPlug : the generic Switch node works perfectly as long as you stay within the confines of the ComputeNode framework, but if you step outside of it you have to accept some additional responsibility. The upshot of this is that it's now essential to call `FilterPlug::sceneAffectsMatch()` rather than `Filter::sceneAffectsMatch()`, so `Filter::sceneAffectsMatch()` is now private.

Breaking Changes :

- GafferScene : Removed FilterSwitch node : use a regular Switch instead. Backwards compatibility
  for old files is provided by converting FilterSwitches during loading.
- Filter : Made `sceneAffectsMatch()` protected. Use the new `FilterPlug::sceneAffectsMatch()` method
  instead.

johnhaddon committed:
GafferImage : Remove ImageSwitch

Breaking Changes :

- GafferImage : Removed ImageSwitch node : use a regular Switch instead. Backwards compatibility
  for old files is provided by converting ImageSwitches during loading.

johnhaddon committed:
Switch : Preconstruct InternedStrings

andrewkaufman committed:
Merge pull request #2810 from medubelko/docsScreenshotFixes

Docs : Fix BuildTarget of two screengrabs

andrewkaufman committed:
Merge pull request #2805 from johnhaddon/removeInputGeneratorCompatibility

Config : Remove InputGenerator compatibility

andrewkaufman committed:
ScriptNode : Use InternedStrings for context access.

andrewkaufman committed:
ScriptNode : Add frameStart and frameEnd context variables.

johnhaddon committed:
Switch : Remove templating, and rename SwitchComputeNode to Switch

A backwards compatibility config file provides compatibility for Python scripts, and typedefs provide source compatibility for C++ files.

Fixes #2123.

johnhaddon committed:
Image/OSL/UI/Scene : Don't rely on SwitchComputeNode typedefs

Refer to the new Switch node directly instead.

johnhaddon committed:
SwitchUI : Improve input nodule spacing

This now matches the spacing for the old ImageSwitch/SceneSwitch nodes.

johnhaddon committed:
FilterPlug : Add `sceneAffectsMatch()` method

This is more convenient than the `Filter::sceneAffectsMatch()` method it replaces, and better fits our general design whereby plugs provide the public interface to the internal capabilities of nodes.

johnhaddon committed:
Replace remaining NodeGraph references with GraphEditor

andrewkaufman committed:
Merge pull request #2813 from johnhaddon/graphEditorRename

Replace remaining NodeGraph references with GraphEditor

johnhaddon committed:
SceneInspector : Respect location selection order

This makes it easier to setup a comparison between two locations (the ordering was pot-luck before, which was confusing).

andrewkaufman committed:
Merge pull request #2814 from johnhaddon/sceneInspectorOrdering

SceneInspector : Respect location selection order

johnhaddon committed:
FilterProcessor : Support non-array inputs in `sceneAffectsMatch()`

johnhaddon committed:
ArrayPlug : Remove FilterPlug compatibility hack

In c113e2f9f9f8137e49c3d3c492aaff2581b9160e we updated `FilterProcessor::sceneAffectsMatch()` to assume that all input plugs were FilterPlugs and not the IntPlugs used prior to version 0.28.0.0. This commit follows suit, turning any attempt to load an obsolete setup into a loading error.

Breaking Changes :

- UnionFilter : Removed compatibility for nodes created prior to version 0.28.0.0.

johnhaddon committed:
BackgroundTask : Add `waitFor()` method

johnhaddon committed:
Revert "SceneGadget : Wait for 100ms before showing progressive updates"

This reverts commit 876af54505aa193bd055f2aa4e21800ca324af1a.

This was a somewhat complex mechanism for batching up 100ms worth of scene update before drawing any change. It also didn't work very well, because during that 100ms further UI interaction could cancel the background update task. For continuous movement such as animation scrubbing or transform manipulation this could lead to continuous cancelling before we showed anything, resulting in a "laggy" feel.

Breaking Change :

- SceneGadget : Removed private member variable. Source compatibility
  is unaffected.

johnhaddon committed:
SceneGadget : Block UI for up to 0.1s when performing updates

This reduces strobing and lag when manipulating things or scrubbing animation.

Commit:
AnimationGadget : snap keys to integer frames

We've had problems with keys ending up slightly off integer frames, which caused
two keys to live right next to each other when the user uses the NodeEditor to
set a keys exactly on an integer frame.

andrewkaufman committed:
Version : Add macro for computing compatibility version

This clarifies client code by allowing:

`#if GAFFER_COMPATIBILITY_VERSION < MAKE_GAFFER_COMPATIBILITY_VERSION( 0, 50 )`

rather than

`#if GAFFER_COMPATIBILITY_VERSION < 50`

andrewkaufman committed:
Merge pull request #2818 from johnhaddon/sceneGadgetLatency

SceneGadget interactivity improvement

johnhaddon committed:
Merge branch 'mattigruener-lightInspector'

This is PR #2797 with some minor tweaks

johnhaddon committed:
GraphEditor : Fix problems with `nodeDoubleClickSignal()`

- The signal didn't respect slot return values, so a slot couldn't return True to signify that it had handled the event, and block other slots.
- The default slot wasn't returning True.

It is now possible to override the double click behaviour in your own configs as follows :

```
GafferUI.GraphEditor.nodeDoubleClickSignal().connect( 0, __myCallback, scoped = False )
```

andrewkaufman committed:
Merge pull request #2821 from johnhaddon/nodeDoubleClickSignalFixes

GraphEditor : Fix problems with `nodeDoubleClickSignal()`

Commit:
AnimationGadget : use std::abs instead of abs and fabs

andrewkaufman committed:
Merge pull request #2811 from andrewkaufman/frameRange

Added script frame range to the context

donboie committed:
GafferSceneUI : SceneInspector - Inform the user if a primvar is indexed

danieldresser committed:
RenderController : Fix bug when object removed from lights set

There is a regression in current Gaffer, introduced in Gaffer 0.48, where if you view a scene where a location is a light, and then view a scene where the same location isn't a light, but has the same object, the object will fail to render.

For example, in this scene:
```
import Gaffer
import GafferScene
import IECore
import imath

__children = {}

__children["Cube"] = GafferScene.Cube( "Cube" )
parent.addChild( __children["Cube"] )
__children["Set"] = GafferScene.Set( "Set" )
parent.addChild( __children["Set"] )
__children["Set"]["in"].setInput( __children["Cube"]["out"] )
__children["Set"]["name"].setValue( '__lights' )
__children["Set"]["paths"].setValue( IECore.StringVectorData( [ '/cube' ] ) )

del __children
```

If you select "Cube", then you see the cube.  If you select "Set" then you will still see the cube, but if you select "Cube" again, it's now disappeared.  In practice, this could happen if a mesh is converted to a mesh light, and then you select a node back up the chain.

The cause is that when /cube is a light, and SceneGraphUpdateTask is run with type = SceneGraph::ObjectType, m_sceneGraph->update() is run with type NoType for /cube ( since it doesn't match the type, since it's considered a light ).  When updateObject() is run with type NoType, it sets m_objectInterface to nullptr, but wasn't clearing m_objectHash, resulting in the null object being reused in the future.

Clearing m_objectHash brings this code more in line with the clearObject() method ( I don't actually fully understand why in this case we're not taking the branch on line 731 which calls clearObject() for things which are clearly not the right type ).  The comparison to clearObject makes me think that this is the right approach, but I really don't fully understand the performance implications of all this code, so we should definitely get John to look at this before merging.

danieldresser committed:
GafferImage::Catalogue : Fix hang when adding image with downsteam network

danieldresser committed:
Moved plug/data conversion from CompoundDataPlug to PlugAlgo

danieldresser committed:
Updated code using PlugAlgo plug/data conversion

danieldresser committed:
Added cameraCompatibility config to deal with old cameras

Commit:
Docs : Add "Anatomy of an Image"

Commit:
"Anatomy of an Image" : Add figures

Commit:
"Anatomy of an Image" : Add indexes

andrewkaufman committed:
Path : Remove ignored qualifier on `isEmpty()`

I'm not sure why this was a const bool return type, but I think just returning bool means the same thing, and it allows us to compile with both -Werror and -Wignored-qualifiers, the later of which comes from the Houdini 17 cxx flags.

andrewkaufman committed:
RendererController : Fix the following compiler errors:

- type qualifiers ignored on function return type
- enumeral and non-enumeral type in conditional expression

andrewkaufman committed:
OSL ShadingEngine : Fix errors from -Werror and -Wmissing-field-initializers

The ClosureDefinition struct and loop over 2 elements seemed unnecessary and allowed me to fix the compilation error without adding extra default constructed ClosureParams.

andrewkaufman committed:
ViewportGadget : Remove ignored qualifier on `getOrthographic3D()`

I'm not sure why this was a const bool return type, but I think just returning bool means the same thing, and it allows us to compile with both -Werror and -Wignored-qualifiers, the later of which comes from the Houdini 17 cxx flags.

johnhaddon committed:
SceneView : Release GIL for `expandSelection()` and `collapseSelection()`

They can trigger background task cancellation which may require the GIL.

johnhaddon committed:
SceneGadget binding : Add missing GIL releases

These functions all lead to background task cancellation, which may require the GIL.

johnhaddon committed:
RenderControllerBinding : Remove dead code and unused includes

johnhaddon committed:
RenderControllerBinding : Add missing GIL releases

These functions all lead to background task cancellation, which may require the GIL.

johnhaddon committed:
Merge pull request #2827 from danieldresser-ie/catalogueHang

GafferImage::Catalogue : Fix hang when adding image with downsteam network

johnhaddon committed:
TransformTool : Support promoted plugs

johnhaddon committed:
BoxUI : Add menu item for promotion of TransformPlugs

They need to be promoted as a whole to be used with the TransformTool.

johnhaddon committed:
Merge pull request #2824 from donboie/sceneInspector

GafferSceneUI : SceneInspector - Inform the user if a primvar is indexed

johnhaddon committed:
Merge pull request #2820 from mattigruener/snappingInAnimationGadget

AnimationGadget : snap keys to integer frames

danieldresser committed:
RenderController : Add test when object removed from lights set

andrewkaufman committed:
IE Config : Allow target apps to override OpenVDB version

andrewkaufman committed:
Merge pull request #2833 from andrewkaufman/ieConfig

IE Config : Allow target apps to override OpenVDB version

danieldresser committed:
GafferArnold : Updated tests for new camera interface

danieldresser committed:
IECoreDelightPreview::Renderer : Adapt to new camera interface

johnhaddon committed:
IECoreDelightPreview::NodeAlgo : Don't use NSIParamIndirect

It no longer exists in the latest 3Delight version (1.1.7 at the time of writing).

johnhaddon committed:
Wrapper : Update for 3Delight 1.1.7

- Use DL_RESOURCES_PATH in place of DL_DISPLAYS_PATH. This seems to be how displays are located now.
- Adjust PYTHONPATH location. NSI bindings are now using `ctypes` to provide a python-version agnostic module.

johnhaddon committed:
Menus : Update for 3delight 1.1.7

johnhaddon committed:
Travis config : Update to 3Delight 1.1.7

johnhaddon committed:
IECoreDelightPreview : Update for camera/screen changes in 3Delight

3Delight has split out the resolution/crop etc into a separate screen node, so the same camera can be rendered at multiple resolutions and so on. CameraAlgo now only deals with creating the camera node, and the Renderer backend deals with the screen node.

andrewkaufman committed:
Merge pull request #2836 from johnhaddon/3delightUpdate

3delight update

danieldresser committed:
GafferDelightTest : Use 3delight's maya lambert as diffuse shader

Previously, it was using a shader named "matte" - I'm not sure where this was supposed to come from?
The 3delight live_edit example uses "matte", but it appears to only ship with this in source form, not compiled.

danieldresser committed:
IECoreDelightPreviewTest : Make previous FOV test actually work

danieldresser committed:
IECoreDelightPreviewTest : Updated for new camera interface

danieldresser committed:
IECoreDelightPreview : Use correct case of "screenwindow"

johnhaddon committed:
Merge pull request #2819 from andrewkaufman/versionMacro

Version : Add macro for computing compatibility version

johnhaddon committed:
Merge pull request #2825 from danieldresser-ie/objectToLight

GafferScene::RenderController : Fix bug when object removed from lights set

johnhaddon committed:
Merge pull request #2830 from johnhaddon/sceneViewGILRelease

GIL management fixes

johnhaddon committed:
Merge pull request #2829 from andrewkaufman/h17compilerErrors

Fixes for Houdini 17 compilation errors

andrewkaufman committed:
Merge pull request #2831 from johnhaddon/transformToolPromotedPlugs

TransformTool : Support promoted plugs

danieldresser committed:
IECoreDelightPreview : Support crop

Figured I might as well add this while I was in here

danieldresser committed:
IECoreAppleseedPreview : Don't convert empty crop at edge into 1 pixel

Previously a crop of (0,1) -> (1,1) would produce a row of pixels, because of clamping the min crop to the max resolution.  Now this will produce an empty region ( consistent with any other crop where min.y = max.y ).

danieldresser committed:
RendererAlgo::Adapt to new camera interface

danieldresser committed:
ImageGadget : Display data window max whenever it is non-default

danieldresser committed:
IECoreArnoldPreviewTest : Remove deprecated from_address calls

Now that AiNodeGetPtr returns node references, we don't use from_address to convert pointers to node references

danieldresser committed:
IECoreArnoldPreviewTest : Arnold 5.2 compat - new address compare

Now that AiNodeGetPtr returns a Python reference object, we need to work a bit more to take the address of the contents to check if two references point to the same node.

johnhaddon committed:
Merge pull request #2840 from danieldresser-ie/dataWindowDisplay

ImageGadget : Display data window max whenever it is non-default

johnhaddon committed:
Update Changes

johnhaddon committed:
ImageView : Fix awkward conditional nesting in `keyPress()`

It doesn't make sense to divide keypresses into those with modifiers and those without. It's clearer to have a single series of if/else clauses.

johnhaddon committed:
ImageView/SceneView : Pause when `Escape` pressed

Fixes #2838

johnhaddon committed:
Travis : Simplify 3Delight config

All this is taken care of by the Gaffer wrapper anyway.

danieldresser committed:
Updated Travis dependencies to version 0.52

andrewkaufman committed:
Merge pull request #2844 from johnhaddon/travisDelightSimplification

Travis : Simplify 3Delight config

andrewkaufman committed:
Merge pull request #2843 from johnhaddon/pauseHotkeys

Pause hotkeys

medubelko committed:
Merge pull request #2832 from medubelko/docsAnatomyOfAnImage

Docs : Add "Anatomy of an Image"

johnhaddon committed:
TaskNode : Fix bugs preventing implementation via internal network

Just as with SceneProcessor and ImageProcessor, we want to allow TaskNode to be subclassed in Python, with the implementation being provided entirely by an internal subnet built in the constructor. This provides the convenience of Python for setup, but without the overhead of actually running Python code during dispatch.

The bug here was that the `preTasks()` and `postTasks()` methods for the internal node were creating tasks not only for the connected external plugs, but also for the unconnected last element of the array. This led to an attempt to create a task for the TaskSubnet node itself, which led to cyclic evaluation. The solution is to only create tasks for output plugs, which we know to be the task outputs from other nodes.

danieldresser committed:
ViewportGadgetTest : Updated for new camera interface

danieldresser committed:
SceneGadgetTest : Update for ViewportGadget changes

andrewkaufman committed:
Merge pull request #2846 from johnhaddon/taskNodeSubnetFix

TaskNode : Fix bugs preventing implementation via internal network

andrewkaufman committed:
Merge pull request #2812 from johnhaddon/switchRefactorRebase3

Switch refactor

danieldresser committed:
SceneGadget::objectAt : If there are multiple hits, pick the nearest

johnhaddon committed:
Merge pull request #2799 from danieldresser-ie/glRenderFixes

OpenGLRender fixes

johnhaddon committed:
Merge pull request #2841 from danieldresser-ie/arnold52back

Support Arnold 5.2 in Tests

johnhaddon committed:
Travis : Update to GafferHQ/dependencies version 0.51.0.0

johnhaddon committed:
TransformTool : Add `selectionChangedSignal()`

Breaking Change :

- TransformTool : Added private member data.

johnhaddon committed:
Bump version to 0.51.0.0 and update Changes

andrewkaufman committed:
Merge pull request #2848 from johnhaddon/selectionChangedSignal

TransformTool : Add `selectionChangedSignal()`

andrewkaufman committed:
Merge pull request #2847 from johnhaddon/51ReleasePrep

Prep for release of version 0.51.0.0

andrewkaufman committed:
IE Config : Add support for variable CXXSTD

andrewkaufman committed:
Merge pull request #2849 from andrewkaufman/ieConfigCXX

IE Config : Add support for variable CXXSTD

andrewkaufman committed:
Bumped major version in preparation for breaking changes

danieldresser-ie committed:
Merge pull request #2816 from danieldresser-ie/cortex10Cam

Camera Overhaul

donboie committed:
GafferUI : VectorDataWidget - prevent tooltips appearing when over data.

donboie committed:
GafferUI : VectorDataWidget - fix horizontal sizing policy.

donboie committed:
GafferUI : VectorDataWidget - remove headerPrefix

donboie committed:
GafferUI : PrimitiveInspector - remove unused variables.
